### ISSUE TITLE
Can this be used as a means to speed up LLM inferencing?

### ISSUE THREAD

**Issue by bulaikexiansheng** [pjlab-sys4nlp/llama-moe#70]
Thanks for your work. I am interested in whether this can be used as a means to speed up LLM reasoning? I have the same idea as your work. I tried to divide `up_proj,`` down_proj, gate_proj` into 4 parts evenly on llama3-8b to form four experts, and added a gate net (dim, num_experts). I try to activate experts with top-2 sparsely, but I found that the reasoning speed will be slower (when not training) than dense model.
My code’s output is showed as below:

```
Once upon a time, in a small village nestled in the rolling hills of the countryside, there lived a young girl named Emily. Emily was a curious and adventurous child, with a heart full of wonder and a mind full of questions. She loved to
Original model forward pass time: 1.6654 seconds
Perplexity: 5.239216327667236

Once upon a time(s gezocht inherited inherited inherited gezocht inherited inherited inherited inherited inherited inherited inherited inherited inherited inherited inherited(s inherited or inherited inheritedSFML inherited or(s inherited inherited inherited inherited or inherited inherited inherited inherited or inherited or inherited inherited inherited(s inherited inherited or inherited
Modified model (Sparse MoE) forward pass time: 2.2228 seconds
Perplexity: 314616.3125
```

**Comment by DaizeDong** [comment]
This method can be used to speed up LLM inferencing with optimization on the hardware. Your observation that MoE lags the speed is also correct, as the MoE forwarding using `nn.ModuleList` is much slower than the dense model due to its dynamic property (e.g. extra cost for dispatching tokens and segmented forwarding). For optimized implementations you can refer to [scattermoe](https://github.com/shawntan/scattermoe) and [MegaBlocks](https://github.com/databricks/megablocks), which can provide faster inference speed for the MoE.

**Comment by bulaikexiansheng** [comment]
Thanks for your answer! I find that if I split FFN into experts, the router's latency can not be ignored. It even exceeds the computation time of sparse experts. The most effective method i tested is random routing. But it just achieve a low speed increase. Thanks for the suggestion, I'll look into it.
